src/uri/punycode.cr

```diff
@@ -0,0 +1,181 @@
+# `Punycode` provides an interface for IDNA encoding (RFC 5980),
+# which is defined in RFC 3493
+#
+# Implementation based on Mathias Bynens `punnycode.js` project
+# https://github.com/bestiejs/punycode.js/
+#
+# RFC 3492:
+# Method to use non-ascii characters as host name of URI
+# https://www.ietf.org/rfc/rfc3492.txt
+#
+# RFC 5980:
+# Internationalized Domain Names in Application
+# https://www.ietf.org/rfc/rfc5980.txt
+class URI
+  class Punycode
+    private BASE         =  36
+    private TMIN         =   1
+    private TMAX         =  26
+    private SKEW         =  38
+    private DAMP         = 700
+    private INITIAL_BIAS =  72
+    private INITIAL_N    = 128
+
+    private DELIMITER = '-'
+
+    private BASE36 = "abcdefghijklmnopqrstuvwxyz0123456789"
+
+    private def self.adapt(delta, numpoints, firsttime)
+      delta /= firsttime ? DAMP : 2
+      delta += delta / numpoints
+      k = 0
+      while delta > ((BASE - TMIN) * TMAX) / 2
+        delta /= BASE - TMIN
+        k += BASE
+      end
+      k + (((BASE - TMIN + 1) * delta) / (delta + SKEW))
+    end
+
+    def self.encode(string)
+      String.build { |io| encode string, io }
+    end
+
+    def self.encode(string, io)
```

Comment:
Just a minor nit bit: methods receiving an optional `IO` should have it prepended as first argument. This makes it easier if maybe additional arguments are added later to keep both signatures similar. Even if that's unlikely I'd recommend to stick with this strategy.

Reply:
`URI.encode` is also accept `String` as first argument and IO as second argument. I think no problem.